addons/account/models/account_move.py

```diff
@@ -1,349 +1,1874 @@
+
+    @api.onchange('invoice_line_ids')
+    def _onchange_invoice_line_ids(self):
+        current_invoice_lines = self.line_ids.filtered(lambda line: not line.exclude_from_invoice_tab)
+        others_lines = self.line_ids - current_invoice_lines
+        if others_lines and current_invoice_lines - self.invoice_line_ids:
+            others_lines[0].recompute_tax_line = True
+        self.line_ids = others_lines + self.invoice_line_ids
+        self._onchange_recompute_dynamic_lines()
+
+    @api.onchange('line_ids', 'invoice_payment_term_id', 'invoice_date_due', 'invoice_cash_rounding_id', 'invoice_vendor_bill_id')
+    def _onchange_recompute_dynamic_lines(self):
+        self._recompute_dynamic_lines()
+
+    @api.model
+    def _build_tax_line_groupby_key(self, line, tax_repartition_line, tax_ids, tag_ids):
+        ''' Create a key used to build a tax mapping between the base lines and the tax lines.
+        This is necessary to aggregate the tax amounts depending of some fields.
+
+        /!\ As line could be a base line or a tax line, only common field could be used to build the key.
+        For example, the tax_ids field could not be used directly from the line as this field has a different
+        purpose on base / tax lines. This is the reason why some field are passed as independent parameters.
+
+        :param line:                    A base / tax account.move.line.
+        :param tax_repartition_line:    An account.tax.repartition.line.
+        :param taxes:                   An account.tax recordset.
+        :param tags:                    An account.tag recordset.
+        :return:                        A list to be hashed representing a key in the taxes_map.
+        '''
+        tax = tax_repartition_line.invoice_tax_id or tax_repartition_line.refund_tax_id
+        return [
+            tax_repartition_line.id,
+            tax_repartition_line.account_id.id or line.account_id.id,
+            line.currency_id.id,
+            tax.analytic and tuple(line.analytic_tag_ids.ids),
+            tax.analytic and line.analytic_account_id.id,
+            tuple(tax_ids),
+            tuple(tag_ids),
+        ]
+
+    @api.model
+    def _get_tax_line_values_from_base_lines(self, base_lines):
+        ''' Hook used to set custom values at the creation of a tax line computed based on base lines.
+        /!\ The result must be consistent with the grouping key (see '_build_tax_line_groupby_key').
+
+        :param base_lines:  A recorset of account.move.line representing the lines on which this tax is computed.
+        :return:            A dictionary of additional values computed from base lines to create a new tax line.
+        '''
+        return {
+            'analytic_account_id': base_lines[0].analytic_account_id.id,
+            'analytic_tag_ids': [(6, 0, base_lines[0].analytic_tag_ids.ids)],
+        }
+
+    @api.multi
+    def _recompute_tax_lines(self, lines_map):
```

Comment:
@antonylesuisse how could one look at this method, understand it, and pretend it's not hackish?